jena-core/src/main/java/com/hp/hpl/jena/reasoner/rulesys/impl/LPBRuleEngine.java

```diff
@@ -52,16 +66,22 @@
     protected boolean recordDerivations;
         
     /** List of engine instances which are still processing queries */
-    protected List<LPInterpreter> activeInterpreters = new ArrayList<>();
-    
+    protected List<LPInterpreter> activeInterpreters = new LinkedList<>();
+
+    protected final int MAX_CACHED_TABLED_GOALS = 
+			Integer.getInteger("jena.rulesys.lp.max_cached_tabled_goals", 512*1024);
+
```

Comment:
getting System properties is problematic in some environments.  See JenaRuntime.getSystemProperty for some mitigation of this.